sbin/mount_tmpfs/mount_tmpfs.8
.\"	$OpenBSD: mount_tmpfs.8,v 1.4 2014/11/16 02:22:10 brad Exp $
.\"	$NetBSD: mount_tmpfs.8,v 1.14 2008/04/30 13:10:53 martin Exp $
.\"
.\" Copyright (c) 2005, 2006 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Julio M. Merino Vidal, developed as part of Google's Summer of Code
.\" 2005 program.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: November 16 2014 $
.Dt MOUNT_TMPFS 8
.Os
.Sh NAME
.Nm mount_tmpfs
.Nd mount an efficient memory file system
.Sh SYNOPSIS
.Nm
.Op Fl g Ar group
.Op Fl m Ar mode
.Op Fl n Ar nodes
.Op Fl o Ar options
.Op Fl s Ar size
.Op Fl u Ar user
.Ar special
.Ar mount_point
.Sh DESCRIPTION
The
.Nm
command attaches an instance of the efficient memory file system to the
global file system namespace.
The file system is populated with data contained by the snapshot pointed to by
.Ar special .
A value of
.Sq tmpfs
results in an empty file system.
The directory specified by
.Ar mount_point
is converted to an absolute path before use and its attributes (owner,
group and mode) are inherited unless explicitly overridden by the options
described below.
.Pp
The following options are supported:
.Bl -tag -width XoXoptions
.It Fl g Ar group
Specifies the group name or GID of the root inode of the file system.
Defaults to the mount point's GID.
.It Fl m Ar mode
Specifies the mode (in octal notation) of the root inode of the file system.
Defaults to the mount point's mode.
.It Fl n Ar nodes
Specifies the maximum number of nodes available to the file system.
If not specified, the file system chooses a reasonable maximum given its
size at mount time, which can be limited with
.Fl s .
.It Fl o Ar options
Options are specified with a
.Fl o
flag followed by a comma-separated string of options.
See the
.Xr mount 8
man page for possible options and their meanings.
.It Fl s Ar size
Specifies the total file system size in bytes.
If zero is given (the default), the available amount of memory (including
main memory and swap space) will be used.
Note that four megabytes are always reserved for the system and cannot
be assigned to the file system.
.It Fl u Ar user
Specifies the user name or UID of the root inode of the file system.
Defaults to the mount point's UID.
.El
.Pp
Every option that accepts a numerical value as its argument can take a
trailing
.Sq b
to indicate bytes (the default), a
.Sq k
to indicate kilobytes, a
.Sq M
to indicate megabytes or a
.Sq G
to indicate gigabytes.
Note that both lowercase and uppercase forms of these letters are allowed.
.Sh EXAMPLES
The following command mounts a tmpfs instance over the
.Pa /tmp
directory, inheriting its owner, group and mode settings:
.Pp
.Dl # mount -t tmpfs tmpfs /tmp
.Pp
The following command mounts a tmpfs instance over the
.Pa /mnt
directory, setting a 20 megabytes limit in space, owned by the
.Sq joe
user and belonging to the
.Sq users
group, with a restricted 0700 mode:
.Pp
.Dl # mount -t tmpfs -o -s20M -o -ujoe -o -gusers -o -m0700 tmpfs /mnt
.Pp
A corresponding
.Xr fstab 5
entry:
.Pp
.Dl tmpfs /mnt tmpfs rw,-s20M,-ujoe,-gusers,-m0700 0 0
.Sh SEE ALSO
.Xr fstab 5 ,
.Xr mount 8
.Sh HISTORY
The
.Nm
utility first appeared in
.Nx 4.0
and
.Ox 5.5 .
Snapshot support first appeared in
.Br 0.1 .
.Sh BUGS
File system meta-data is not pageable.
If there is not enough main memory to hold this information, the system may
become unstable or very unresponsive because it will not be able to allocate
required memory.
A malicious user could trigger this condition if he could create lots of
files inside a size-unbounded tmpfs file system.
Limiting the number of nodes per file system
.Pq Fl n
will prevent this; the default value for this setting is also often adjusted
to an adequate value to resolve this.


sbin/mount_tmpfs/mount_tmpfs.c
/*	$NetBSD: mount_tmpfs.c,v 1.24 2008/08/05 20:57:45 pooka Exp $	*/

/*
 * Copyright (c) 2005, 2006 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Julio M. Merino Vidal, developed as part of Google's Summer of Code
 * 2005 program.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if 0
__RCSID("$NetBSD: mount_tmpfs.c,v 1.24 2008/08/05 20:57:45 pooka Exp $");
#endif

#include <sys/param.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <grp.h>
#include <inttypes.h>
#include <mntopts.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <util.h>

#include "mount_tmpfs.h"

/* --------------------------------------------------------------------- */

static const struct mntopt mopts[] = {
	MOPT_STDOPTS,
	MOPT_UPDATE,
	MOPT_SNAPSHOT,
	{ NULL },
};

/* --------------------------------------------------------------------- */

static void	usage(void) __dead;
static uid_t	a_uid(const char *);
static gid_t	a_gid(const char *);
static uid_t	a_gid(const char *);
static int	a_num(const char *, const char *);
static mode_t	a_mask(const char *);
static void	pathadj(const char *, char *);

/* --------------------------------------------------------------------- */

void
mount_tmpfs_parseargs(int argc, char *argv[],
	struct tmpfs_args *args, int *mntflags,
	char *canon_dev, char *canon_dir)
{
	int gidset, modeset, uidset; /* Ought to be 'bool'. */
	int ch;
	gid_t gid;
	uid_t uid;
	mode_t mode;
	int64_t tmpnumber;
	struct stat sb;

	/* Set default values for mount point arguments. */
	memset(args, 0, sizeof(*args));
	args->ta_version = TMPFS_ARGS_VERSION;
	args->ta_size_max = 0;
	args->ta_nodes_max = 0;
	*mntflags = 0;

	gidset = 0; gid = 0;
	uidset = 0; uid = 0;
	modeset = 0; mode = 0;

	optind = optreset = 1;
	while ((ch = getopt(argc, argv, "g:m:n:o:s:u:")) != -1 ) {
		switch (ch) {
		case 'g':
			gid = a_gid(optarg);
			gidset = 1;
			break;

		case 'm':
			mode = a_mask(optarg);
			modeset = 1;
			break;

		case 'n':
			if (scan_scaled(optarg, &tmpnumber) == -1)
				err(EXIT_FAILURE, "failed to parse nodes `%s'",
				    optarg);
			args->ta_nodes_max = tmpnumber;
			break;

		case 'o':
			getmntopts(optarg, mopts, mntflags);
			break;

		case 's':
			if (scan_scaled(optarg, &tmpnumber) == -1)
				err(EXIT_FAILURE, "failed to parse size `%s'",
				    optarg);
			args->ta_size_max = tmpnumber;
			break;

		case 'u':
			uid = a_uid(optarg);
			uidset = 1;
			break;

		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 2)
		usage();

	if (!strcmp(argv[0], "tmpfs"))
		strlcpy(canon_dev, argv[0], MAXPATHLEN);
	else
		pathadj(argv[0], canon_dev);
	pathadj(argv[1], canon_dir);

	if (stat(canon_dir, &sb) == -1)
		err(EXIT_FAILURE, "cannot stat `%s'", canon_dir);

	args->ta_root_uid = uidset ? uid : sb.st_uid;
	args->ta_root_gid = gidset ? gid : sb.st_gid;
	args->ta_root_mode = modeset ? mode : sb.st_mode;
	args->ta_fspec = canon_dev;
}

/* --------------------------------------------------------------------- */

static void
usage(void)
{
	extern char *__progname;
	(void)fprintf(stderr,
	    "usage: %s [-g group] [-m mode] [-n nodes] [-o options] [-s size]\n"
	    "           [-u user] tmpfs mount_point\n", __progname);
	exit(1);
}

/* --------------------------------------------------------------------- */

int
mount_tmpfs(int argc, char *argv[])
{
	struct tmpfs_args args;
	char canon_dev[MAXPATHLEN], canon_dir[MAXPATHLEN];
	int mntflags;

	mount_tmpfs_parseargs(argc, argv, &args, &mntflags,
	    canon_dev, canon_dir);

	if (mount(MOUNT_TMPFS, canon_dir, mntflags, &args) == -1)
		err(EXIT_FAILURE, "%s on %s", canon_dev, canon_dir);

	return EXIT_SUCCESS;
}

int
main(int argc, char *argv[])
{

	/* setprogname(argv[0]); */
	return mount_tmpfs(argc, argv);
}

static uid_t
a_uid(const char *s)
{
	struct passwd *pw;

	if ((pw = getpwnam(s)) != NULL)
		return pw->pw_uid;
	return a_num(s, "user");
}

static gid_t
a_gid(const char *s)
{
	struct group *gr;

	if ((gr = getgrnam(s)) != NULL)
		return gr->gr_gid;
	return a_num(s, "group");
}

static int
a_num(const char *s, const char *id_type)
{
	int id;
	char *ep;

	id = strtol(s, &ep, 0);
	if (*ep || s == ep || id < 0)
		errx(1, "unknown %s id: %s", id_type, s);
	return id;
}

static mode_t
a_mask(const char *s)
{
	int rv;
	char *ep;

	rv = strtol(s, &ep, 8);
	if (s == ep || *ep || rv < 0)
		errx(1, "invalid file mode: %s", s);
	return rv;
}

static void
pathadj(const char *input, char *adjusted)
{

	if (realpath(input, adjusted) == NULL)
		err(1, "realpath %s", input);
}
